text/0000-multi-type-return-position-impl-trait.md

```diff
@@ -0,0 +1,356 @@
+[TAIT]: https://rust-lang.github.io/rfcs/2515-type_alias_impl_trait.html
+
+A motivating example for this is use in error handling: it's not uncommon to
+have a function return more than one error type, but you may not necessarily
+care about the exact errors returned. You may either choose to define a `Box<dyn
+Error + 'static>` which has the downside that [it itself does not implement
+`Error`][no-error]. Or you may choose to define your own enum of errors, which
+can be a lot of work and may obfuscate the actual intent of the code. It may
+sometimes be preferable to return an `impl Trait` instead:
+
+[no-error]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=97894fc907fa2d292cbe909467d4db4b
+
+```rust
+use std::error::Error;
+use std::fs;
+
+// ❌ Multi-type RPIT does not yet compile (Rust 1.66.0)
+// error[E0282]: type annotations needed
+fn main() -> Result<(), impl Error> {
+    let num = i8::from_str_radix("A", 16)?;       // `Result<_, std::num::ParseIntError>`
+    let file = fs::read_to_string("./file.csv")?; // `Result<_, std::io::Error>`
+    // ... use values here
+    Ok(())
+}
+```
+
+# Desugaring
+[reference-level-explanation]: #reference-level-explanation
+
+## Overview
+
+Let's take a look again at the code from our motivation section. This function
+has two branches which each return a different type which implements the
+[`Iterator` trait][`Iterator`]:
+
+[`Iterator`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html
+
+```rust
+fn multi_iter(x: i32) -> impl Iterator<Item = i32> {
+    match x {
+        0 => 1..10,                   // `std::ops::Range<i32>`
+        _ => vec![5, 10].into_iter(), // `std::vec::IntoIter<i32>`
+    }
+}
+```
+
+This code should be desugared by the compiler into something resembling the following
+([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af4c0e61df25acaada168449df9838d3)):
+
+```rust
+// anonymous enum generated by the compiler
+enum Enum {
+    A(std::ops::Range<i32>),
+    B(std::vec::IntoIter<i32>),
+}
+
+// trait implementation generated by the compiler,
+// delegates to underlying enum member's values
+impl Iterator for Enum {
+    type Item = i32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Enum::A(iter) => iter.next(),
+            Enum::B(iter) => iter.next(),
+        }
+    }
+
+    // ..repeat for the remaining 74 `Iterator` trait methods
+}
+
+// the desugared function now returns the generated enum
+fn multi_iter(x: i32) -> Enum {
+    match x {
+        0 => Enum::A(1..10),
+        _ => Enum::B(vec![5, 10].into_iter()),
+    }
+}
+```
+
+## Step-by-step guide
+
+This desugaring can be implemented using the following steps:
+
+1. Find all return calls in the function
+2. Define a new enum with a member for each of the function's return types
+3. Implement the traits declared in the `-> impl Trait` bound for the new enum,
```

Comment:
How do you do this for traits that have fn() -> Self? For example, Default doesn't have an obvious delegation.

This seems like a key thing to discuss, particularly as it means that the addition of such methods might be semver-breaking, even in a sealed trait.

Reply:
> How do you do this for traits that have fn() -> Self? For example, Default doesn't have an obvious delegation.

Well, maybe that's just one of the restrictions that may have to come with this feature.  
I think returning `Self` could work if you're method chaining. For example `SomeType::impl_fn().do_something().get_concrete_value();`. Though, not being able to use any `-> Self` methods wouldn't be too big of a deal breaker if it meant we had this feature.

Reply:
To: @Mark-Simulacrum, @CAD97
CC: @yoshuawuyts, @Ekleog, @glaebhoerl

I thoughts of a really interesting solution regarding `trait`s that have `fn() -> Self` (or receiver variants).  
Basically, calling a method that returns `Self` is an error because there is we can't know what that type is, we only know what method that value has that we can call. - I believe [E0119](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/error-index.html#E0119) and [E0411](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/error-index.html#E0411) are both already compiler errors that could possibly represent this.  
It could return a simple compiler error like:
```
/* Human style. */
Can not return ambiguous `Self` for static dispatch type `impl Trait`.


/* Structured style. */
error: ambiguous type `Self` on associated method return type
```
Regardless of whether this feature would need its own error code or falls under something like [E0411](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/error-index.html#E0411) I've decided to make examples of various different situations.


## Examples of Theoretical Errors

### Attempted type-inference for static-dispatch-type `{impl T}`.
Erroneous code:
```rust
trait Baz {
  fn buzz(&self) {}
}

struct Foo;
impl Baz for Foo {}
struct Bar;
impl Baz for Bar {}

fn some_value() -> impl Baz {
  /* ... */
}

fn main() {
  let foo = some_value(); // error: can not infer single type for `foo` in the set of
                          //        all types that implement `Baz`.
}
```
To fix the code the compiler may suggest calling the `.buzz()` method of `Baz`.  
I.e:
```rust
trait Baz {
  fn buzz(&self) -> &str { "Buzz!" }
}

struct Foo;
impl Baz for Foo {}
struct Bar;
impl Baz for Bar {}

fn some_value() -> impl Baz {
  /* ... */
}

fn main() {
  let buzz = some_value().buzz(); // this is o.k. because we know we'll
                                  // always get an `&str` from `{impl Baz}::buzz(&self)`.
}
```

### Unresolved type `Self` for `{impl T}::name() -> Self`.
Like with the previous error, the crux of the issue is inferencing and assigning a type for `Self`, which I don't believe is possible.

Erroneous code:
```rust
trait Baz {
  fn fizz(&self) -> Self {/* ... */}
}

struct Foo;
impl Baz for Foo {}
struct Bar;
impl Baz for Bar {}

fn some_value() -> impl Baz {/* ... */}

fn main() {
  let fizzed = some_value().fizz(); // error: can not infer the type of `Self` for.
                                    //        local value of type `{impl Baz}`.
}
```

## Explicitness
As suggested, an explicit syntax like `enum impl T`, which would turn:
```rust
fn some_value() -> impl MyTrait {/* ... */}
```
into:
```rust
fn some_value() -> enum impl MyTrait {/* ... */}
```
It would also be logical to accordingly adjust the "`{impl T}`" placeholder into "`{enum impl T}`", but I think the former suffices.  
With this explicitness it makes it easy to pinpoint functions that could be potential troublemakers for the kinds of scenarios we can think up. It also allows us to keep the normal `-> impl T` behavior, and extend that functionality to static dispatching.